File: Lintcode 013 - strStr.java
Lintcode - 013. strStr.java

class Solution {
    /**
     * Returns a index to the first occurrence of target in source,
     * or -1  if target is not part of source.
     * @param source string to be scanned.
     * @param target string containing the sequence of characters to match.
     */
    public int strStr(String source, String target) {
        //write your code here
        if (source == null || target == null) {
            return -1;
        }
        
        int i, j;
        //注意，source的循环只到source.length() - target.length() + 1
        for(i = 0; i < source.length() - target.length() + 1; i++) {
            for (j = 0; j < target.length(); j++) {
                if(source.charAt(i + j) != target.charAt(j)){
                    break;
                    //break，跳出整个 for j 循环
                    //continue，继续下一个 ++ 循环
                }
            }
            //注意这个 if j 判断放在for j 结束后
            if (j == target.length()){
                return i;
            }
        }
        return -1;
    }
}
 

File: Lintcode 013 - strStr.cs
using System;
class Solution
{
    static void Main()
    {
        string source = "qwerty";
        string target = "rty";
        
        int result = strStr(source, target);
        Console.WriteLine(result + "");
    
    }
    
    static int strStr(string source, string target)
    {
        if(source == null || source.Length == 0) 
        {
            return -1;
        }
        
        int i;
        int j;
        for (i = 0; i < source.Length + target.Length - 1; i++) 
        {
            for (j = 0; j < target.Length; j++)
            {
                if (source[i + j] != target[j])
                {
                    break;
                }
            }
            
            if(j == target.Length)
            {
                return i;    
            }
        }              
        return -1;
    }
}
